cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -805,6 +805,7 @@ func startServiceAccountTokenController(ctx context.Context, controllerContext C
 		return nil, false, fmt.Errorf("failed to build token generator: %v", err)
 	}
 	tokenController, err := serviceaccountcontroller.NewTokensController(
+		ctx,
```

Comment:
I believe we discussed that we should not require context for the controller creation, only for the actual run method. If you change this to only pass logger, similarly to what the change below was before your modification we could also get rid of the changes in the controller to use `.AddEventHandler` for now. This would also address part of the concerns that Jordan had in [this thread](https://github.com/kubernetes/kubernetes/pull/126387#discussion_r1808956552).

Reply:
Ack (https://github.com/kubernetes/kubernetes/pull/126387#discussion_r1879443708).

Reply:
@soltysh: so do you also want to avoid adding `AddEventHandlerWithContext`? Or add `AddEventHandlerWithLogger` instead?

The problem is that there's a bunch of log calls potentially triggered by `AddEventHandler`. Without something passed by the caller, those cannot be properly attributed to the caller. In particular when those log calls point out things like "Warning: resync period is smaller than resync check period and the informer has already started" it would be good to get more information into the log entry from a contextual logger.

Reply:
Coming back to `*WithContext`: part of my motivation for adding it was that it's easier for developers to cancel one context and get everything stopped which was created with it than having to remember to call all individual `Stop` or `Remove` methods. Is that goal worth extra complexity?

Perhaps not. Resource leaks are annoying, but perhaps developers simply have to be more careful if they care.

> In particular when those log calls point out things like "Warning: resync period is smaller than resync check period and the informer has already started" it would be good to get more information into the log entry from a contextual logger.

That may be possible without an API change. The code underneath `AddEventHandler` can mark itself as helper code, then stack unwinding will skip it and the log message will be recorded as coming from the code which made the broken `AddEventHandler` call. How does that sound?

Reply:
I removed AddEventHandlerWithContext and then my own new `TestEventPanics` reminded me why a per-event-handler logger is useful: when it panics, we want to know what panicked. I've replaced `AddEventHandlerWithContext` with `AddEventHandlerWithOptions`, with a logger as one optional option.

Reply:
> I've replaced `AddEventHandlerWithContext` with `AddEventHandlerWithOptions`, with a logger as one optional option.

This will probably serve us much better in the long run, w/o the need to change the API.